shell/browser/api/atom_api_url_loader.cc

```diff
@@ -0,0 +1,396 @@
+  gin::ObjectTemplateBuilder GetObjectTemplateBuilder(
+      v8::Isolate* isolate) override {
+    return gin::Wrappable<JSChunkedDataPipeGetter>::GetObjectTemplateBuilder(
+               isolate)
+        .SetMethod("write", &JSChunkedDataPipeGetter::WriteChunk)
+        .SetMethod("done", &JSChunkedDataPipeGetter::Done);
+  }
+
+  static gin::WrapperInfo kWrapperInfo;
+  ~JSChunkedDataPipeGetter() override = default;
+
+ private:
+  JSChunkedDataPipeGetter(
+      v8::Isolate* isolate,
+      v8::Local<v8::Function> body_func,
+      mojo::PendingReceiver<network::mojom::ChunkedDataPipeGetter>
+          chunked_data_pipe_getter)
+      : isolate_(isolate), body_func_(isolate, body_func) {
+    receiver_.Bind(std::move(chunked_data_pipe_getter));
+  }
+
+  // network::mojom::ChunkedDataPipeGetter:
+  void GetSize(GetSizeCallback callback) override {
+    size_callback_ = std::move(callback);
+  }
+
+  void StartReading(mojo::ScopedDataPipeProducerHandle pipe) override {
+    data_producer_ = std::make_unique<mojo::DataPipeProducer>(std::move(pipe));
+    DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+    v8::HandleScope handle_scope(isolate_);
+    v8::MicrotasksScope script_scope(isolate_,
+                                     v8::MicrotasksScope::kRunMicrotasks);
+    auto maybe_wrapper = GetWrapper(isolate_);
+    v8::Local<v8::Value> wrapper;
+    if (!maybe_wrapper.ToLocal(&wrapper)) {
+      return;
+      // hopefully this just drops the pipe and we're good?
+    }
+    v8::Local<v8::Value> argv[] = {wrapper};
+    node::Environment* env = node::Environment::GetCurrent(isolate_);
+    auto global = env->context()->Global();
+    node::MakeCallback(isolate_, global, body_func_.Get(isolate_),
+                       node::arraysize(argv), argv, {0, 0});
+  }
+
+  v8::Local<v8::Promise> WriteChunk(v8::Local<v8::Value> buffer_val) {
+    gin_helper::Promise<void> promise(isolate_);
+    v8::Local<v8::Promise> handle = promise.GetHandle();
+    if (!buffer_val->IsArrayBufferView()) {
+      promise.RejectWithErrorMessage("Expected an ArrayBufferView");
+      return handle;
+    }
+    if (is_writing_) {
+      promise.RejectWithErrorMessage("Only one write can be pending at a time");
+      return handle;
+    }
+    if (!size_callback_) {
+      promise.RejectWithErrorMessage("Can't write after calling done()");
+      return handle;
+    }
+    auto buffer = buffer_val.As<v8::ArrayBufferView>();
+    is_writing_ = true;
+    bytes_written_ += buffer->ByteLength();
+    auto backing_store = buffer->Buffer()->GetBackingStore();
+    data_producer_->Write(
+        std::make_unique<BufferDataSource>(base::make_span(
+            static_cast<char*>(backing_store->Data()) + buffer->ByteOffset(),
+            buffer->ByteLength())),
+        base::BindOnce(
+            &JSChunkedDataPipeGetter::OnWriteChunkComplete,
+            base::Unretained(
+                this),  // TODO: should this be a weak ptr? what do we do with
```

Comment:
We can probably add an option to `gin_helper::Promise`, to make it throw in the C++ destructor if it has not been resolved or thrown before. So if `JSChunkedDataPipeGetter` goes away in the middle of writing, the returned JS promise would throw, other than pending forever.

It should probably even be the default behavior.

Reply:
Ah, yes, that would make sense!

Reply:
Actually... I'm not so sure this is a good idea. It would mean entering JS at a possibly unexpected time, when it's not necessarily safe to begin executing JS. Probably best that it remains explicit.